### PR TITLE
Generalize percussion set architecture

### PR DESCRIPTION
**Pull request by danielafeitosa**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -31,9 +31,27 @@
 // 74 - Soft Tek Front
 
 let drums = null;
-async function loadInstrument() {
+let buttons = null;
+const config = {}
+
+async function loadSoundFonts(instrument) {
+  let soundFont = await fetch('/soundfonts.json')
```
Comment (infojunkie) [inline app.js]: use `const` if the variable won't be mutated

File: app.js [hunk 2]
```diff
@@ -31,9 +31,27 @@
...
+  if (!config[instrument]) {
+    config[instrument] = await loadSoundFonts(instrument);
+  }
+  return config[instrument];
+}
+
+async function loadInstrument(instrument) {
   return new Promise((resolve, reject) => {
-    Soundfont.instrument(new AudioContext(), 'doumbek', { soundfont: 'Doumbek-Faisal', nameToUrl: function(name, soundfont, format) {
+    let soundfont = config[instrument].soundfont
```
Comment (infojunkie) [inline app.js]: `const`

File: app.js [hunk 3]
```diff
@@ -44,18 +62,15 @@ async function loadInstrument() {
 }
 
 async function playButton(button) {
+  let instrument = 'doumbek';
```
Comment (infojunkie) [inline app.js]: Better hardcode default values as a global `const DEFAULT_INSTRUMENT`. Note you also probably need a `DEFAULT_SOUNDFONT`.

File: soundfonts.json [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (infojunkie) [inline soundfonts.json:1]: This file should be organized as soundfonts => instruments => mapping of MIDI keys to percussion names e.g. https://github.com/infojunkie/drumkit/blob/a95f4d242c515454e50bb778052a4246d7d8ef4d/app.js#L6-L15

The mapping of input buttons to specific MIDI keys should be kept dynamic, and defaulted with something like `const DEFAULT_BUTTONS`.

**Comment by danielafeitosa** [comment]
@infojunkie updated the PR with the changes requested, please review it again :blush: 

I set the `DEFAULT_BUTTONS` with the last 6 sound codes but then I map the buttons with the first 6 codes from soundfonts. I did it to test that the codes were being updated. Let me know if I should initialize the `DEFAULT_BUTTONS` with the first 6 values

File: app.js [hunk 5]
```diff
@@ -31,9 +31,36 @@
...
+};
+
+async function loadSoundFonts(soundfont) {
+  const soundFonts = await fetch('/soundfonts.json')
+    .then(data => data.json())
+    .then(json => json)
+  return soundFonts[soundfont]
+}
+
+async function loadConfig(soundfont, instrument) {
```
Comment (infojunkie) [inline app.js]: I think the functions `loadSoundFonts` and `loadConfig` should be a single one that just loads the `soundfont.json` file and places it in `config`. We can then call this function when the script loads, and any subsequent function that needs soundfont information just reads from `config`.

File: app.js [hunk 6]
```diff
@@ -31,9 +31,36 @@
...
+  if (!config[soundfont]) {
+    config[soundfont] = await loadSoundFonts(soundfont);
+  }
+  return config[soundfont][instrument];
+}
+
 async function loadInstrument() {
   return new Promise((resolve, reject) => {
-    Soundfont.instrument(new AudioContext(), 'doumbek', { soundfont: 'Doumbek-Faisal', nameToUrl: function(name, soundfont, format) {
+    Soundfont.instrument(new AudioContext(), DEFAULT_INSTRUMENT, { soundfont: DEFAULT_SOUNDFONT, nameToUrl: function(name, soundfont, format) {
```
Comment (infojunkie) [inline app.js]: It does not make sense to use the defaults in this function. We can define global `soundfont` and `instrument` variables that represent the current selection, and that are initialized to the defaults.

File: app.js [hunk 7]
```diff
@@ -31,9 +31,36 @@
 // 74 - Soft Tek Front
 
 let drums = null;
+let buttons = null;
```
Comment (infojunkie) [inline app.js]: I think this should be initialized after `DEFAULT_BUTTONS` and set as a copy of it.

File: app.js [hunk 8]
```diff
@@ -43,19 +70,26 @@ async function loadInstrument() {
   });
 }
 
+function mapButtons(availableButtons) {
```
Comment (infojunkie) [inline app.js]: If `buttons` is set to a copy of `DEFAULT_BUTTONS`, then we can defer the mapping from buttons to instrument sounds to some future UI, and this function is no longer needed.

File: app.js [hunk 9]
```diff
@@ -43,19 +70,26 @@ async function loadInstrument() {
...
     drums = await loadInstrument();
   }
-  const buttonToSound = {
-    'b1': 60,
-    'b2': 61,
-    'b3': 62,
-    'b4': 63,
-    'b5': 64,
-    'b6': 65,
+  let percussionSet = await loadConfig(DEFAULT_SOUNDFONT, DEFAULT_INSTRUMENT);
```
Comment (infojunkie) [inline app.js]: These 4 lines are unneeded if we make the changes outlined above. Essentially, this function only does the following:
- loads the drums if not loaded
- plays the sound corresponding to the already mapped button

File: app.js [hunk 10]
```diff
@@ -31,9 +31,36 @@
 // 74 - Soft Tek Front
```
Comment (infojunkie) [inline app.js]: We can remove the Doumbek comments since we moved the relevant info to `soundfonts.json`.

**Comment by infojunkie** [comment]
Thanks for these changes :heart: Left some more comments :sweat_smile:

**Pull request merged**